### ISSUE TITLE
azurerm_function_app_host_keys not working

### ISSUE THREAD

**Issue by julikru** [hashicorp/terraform-provider-azurerm#16599]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

1.0.0

### AzureRM Provider Version

3.3.0

### Affected Resource(s)/Data Source(s)

azurerm_function_app_host_keys

### Terraform Configuration Files

```hcl
resource "azurerm_linux_function_app" "core" {
  name                       = "coreapi_name"
  location                   = "westeurope"
  resource_group_name        = "your-rg"
  service_plan_id        = azurerm_service_plan.this.id
  storage_account_name       = azurerm_storage_account.apps.name
  storage_account_access_key = azurerm_storage_account.apps.primary_access_key

  site_config {
    pre_warmed_instance_count = 1
  }

  app_settings = {
...
}

  identity {
    type = "SystemAssigned"
  }

  tags = {
    environment = "${var.env_code}"
  }
}

data "azurerm_function_app_host_keys" "this" {
  name                  =  azurerm_linux_function_app.core.name
  resource_group_name   =  data.azurerm_resource_group.this.name
}
```

### Debug Output/Panic Output

```shell
ata.azurerm_function_app_host_keys.this: Still reading... [19m21s elapsed]
data.azurerm_function_app_host_keys.this: Still reading... [19m31s elapsed]
data.azurerm_function_app_host_keys.this: Still reading... [19m41s elapsed]
data.azurerm_function_app_host_keys.this: Still reading... [19m51s elapsed]
data.azurerm_function_app_host_keys.this: Still reading... [20m1s elapsed]
╷
│ Error: making Read request on Function App: (Site Name "coreapi-prod" / Resource Group "RG-XX"): web.AppsClient#ListHostKeys: Failure sending request: StatusCode=0 -- Original Error: context deadline exceeded
│ 
│   with data.azurerm_function_app_host_keys.this,
│   on asp.tf line 84, in data "azurerm_function_app_host_keys" "this":
│   84: data "azurerm_function_app_host_keys" "this" {
│
```

### Expected Behaviour

The host keys should be fetched and it should be possible to use it with azurerm_function_app_host_keys.this.default_function_key

### Actual Behaviour

Timeout, reproduced 10 times

### Steps to Reproduce

terraform apply

**Comment by DSollickOnyx** [comment]
Hey @julikru I've run into this as well -- it's not possible (at least as of writing) to get the function app key before the function app has actually been deployed via CLI or what-have-you. I agree it does not work as one would hope. Issue #8515 deals with this as well and points up to a core Terraform bug [26074](https://github.com/hashicorp/terraform/issues/26074).

**Comment by DSollickOnyx** [comment]
@JonathonAnderson Right! I meant like actually deploying the code/package/container what-have-you. Definitely agreed that TF should and could handle this situation more gracefully, but at the core it seems to be Azure being inconsistent. We've gotten around this by doing the TF deployment, then the code deployment, and then back into TF but that's obviously a pain.

**Comment by JonathonAnderson** [comment]
@DSollickOnyx Is key generation related to app deployment? I've used this same plan in multiple times in multiple regions because we've had issues with quotas and service availability, and most of the time, the function apps deploy, and the keys are generated, but we still have not deployed our apps. So, this issue doesn't seem entirely related to code/package/container deployment. In some of the cases, 25% maybe, the keys are not created and the process times out.

**Comment by DSollickOnyx** [comment]
@JonathonAnderson I haven't necessarily experimented with it enough to give you a 100% solid answer, but if Azure is creating the keys lazily it's possible that deploying code to it forces that to happen? I've never had an issue getting function keys to a function app that's been deployed **to** whereas an empty function app is hit-or-miss as you've said. Googling around people have had all sorts of reasons why the keys don't exist from storage account issues to having deleted a function app with the same name in the last few hours. It could even just be that the delay from waiting for the code to deployed is long enough for them to be generated and available to TF?

**Comment by JonathonAnderson** [comment]
@DSollickOnyx Okay, that's interesting. I didn't find those results, but I was searching specifically for Terraform issues, not general Function App issues. I, unfortunately, assumed this was a Terraform problem. In this case, I had just deleted the entire App Service Plan that they belonged to and the Function Apps along with it, so they were recreated within an hour of being deleted. I created the keys manually and Terraform completed.

**Comment by scott-doyland-burrows** [comment]
If the data source ```azurerm_function_app_host_keys``` can obtain the keys, then can the ```azurerm_linux_function_app``` just be updated to export them.

We shouldn't really need to use a data source to obtain attributes from a resource that is created within the same Terraform run.

**Comment by StefanSchoof** [comment]
I think we found a cause and a solution:
We set the `WEBSITE_RUN_FROM_PACKAGE` by the creation to 1 of the function results into no function key via tf (or in the portal). Our solution is now to not set this via terraform and ignore it.
```terraform
  lifecycle {
    prevent_destroy = true
    ignore_changes = [
      app_settings["WEBSITE_RUN_FROM_PACKAGE"] # Adding this to avoid deletion of functions which are linked as "run from package"
    ]
  }
```
It is set by the first deployment via the pipeline.

**Comment by ChristianSauer** [comment]
Funny, I revisited an old project of mine, googled this error and found my old bug report from 2020
https://github.com/hashicorp/terraform-provider-azurerm/issues/9869

But hey, closed because it's "stale"...

**Comment by sbx0r** [comment]
I have the same issue.
Try to play around with different Azure Service Plan tiers. Changing from B2 -> B3 helped in my case

**Comment by sbx0r** [comment]
In my case the solution was related to recreating a function app with the same name using the same storage account (https://learn.microsoft.com/en-us/azure/azure-functions/errors-diagnostics/diagnostic-events/azfd0007). So I wiped out the blob and it worked

**Comment by tullydwyer** [comment]
> @JonathonAnderson I haven't necessarily experimented with it enough to give you a 100% solid answer, but if Azure is creating the keys lazily it's possible that deploying code to it forces that to happen? I've never had an issue getting function keys to a function app that's been deployed **to** whereas an empty function app is hit-or-miss as you've said. Googling around people have had all sorts of reasons why the keys don't exist from storage account issues to having deleted a function app with the same name in the last few hours. It could even just be that the delay from waiting for the code to deployed is long enough for them to be generated and available to TF?

For anyone having issue that the key is missing, from my my research the key is only generated on code deployment: https://github.com/Azure/azure-cli/issues/23693

**Comment by michasacuer** [comment]
Still same issue with azurerm_windows_function_app, host keys are not created via terraform. Provider is the latest, 4.9.0. Is there any workaround?
